llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/metrics/LlapMetricsCollector.java

```diff
@@ -0,0 +1,194 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Collect metrics from the llap daemons in every given milliseconds.
+ */
+public class LlapMetricsCollector implements ServiceInstanceStateChangeListener<LlapServiceInstance> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LlapMetricsCollector.class);
+  private static final String THREAD_NAME = "LlapTaskSchedulerMetricsCollectorThread";
+  private static final long INITIAL_DELAY_MSEC = 60000L;
+
+  private final Configuration conf;
+  private final ScheduledExecutorService scheduledMetricsExecutor;
+  private final LlapManagementProtocolClientImplFactory clientFactory;
+  private final Map<String, LlapManagementProtocolClientImpl> llapClients;
+  private final Map<String, LlapMetrics> instanceStatisticsMap;
+  private final long metricsCollectionMs;
+
+
+  public LlapMetricsCollector(Configuration conf) {
+    this(
+            conf,
+            Executors.newSingleThreadScheduledExecutor(
+                    new ThreadFactoryBuilder().setDaemon(true).setNameFormat(THREAD_NAME)
+                            .build()),
+            LlapManagementProtocolClientImplFactory.basicInstance(conf));
+  }
+
+  @VisibleForTesting
+  LlapMetricsCollector(Configuration conf, ScheduledExecutorService scheduledMetricsExecutor,
+                       LlapManagementProtocolClientImplFactory clientFactory) {
+    this.conf = conf;
+    this.scheduledMetricsExecutor = scheduledMetricsExecutor;
+    this.clientFactory = clientFactory;
+    this.llapClients = new HashMap<>();
+    this.instanceStatisticsMap = new ConcurrentHashMap<>();
+    this.metricsCollectionMs = HiveConf.getTimeVar(conf,
+            HiveConf.ConfVars.LLAP_TASK_SCHEDULER_AM_COLLECT_DAEMON_METRICS_MS, TimeUnit.MILLISECONDS);
+  }
+
+  public void start() {
+    if (enabled()) {
+      scheduledMetricsExecutor.scheduleAtFixedRate(() -> {
+        collectMetrics();
+      }, INITIAL_DELAY_MSEC, metricsCollectionMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private boolean enabled() {
+    return metricsCollectionMs > 0;
+  }
+
+  public void shutdown() {
+    scheduledMetricsExecutor.shutdownNow();
+  }
+
+  @VisibleForTesting
+  void collectMetrics() {
+    if (enabled()) {
+      for (Map.Entry<String, LlapManagementProtocolClientImpl> entry : llapClients.entrySet()) {
+        String identity = entry.getKey();
+        LlapManagementProtocolClientImpl client = entry.getValue();
+        try {
+          LlapDaemonProtocolProtos.GetDaemonMetricsResponseProto metrics =
+                  client.getDaemonMetrics(null,
+                          LlapDaemonProtocolProtos.GetDaemonMetricsRequestProto.newBuilder().build());
+          instanceStatisticsMap.put(identity, new LlapMetrics(metrics));
+
+        } catch (ServiceException ex) {
+          LOG.error(ex.getMessage(), ex);
```

Comment:
Should we maybe remove the previously stored metrics from instanceStatisticsMap? I know, we have an associated timestamp there but it would make the code "safer" if we don't store old metrics whenever an error happens when we try to receive new daemon metrics.

Reply:
Done.